### PR TITLE
Add type hints to our public API  (PEPs 484 & 561)

### LINKED ISSUE #200: Add static typing to Hypothesis API
The new typing module in 3.5 and [PEP 0484](https://www.python.org/dev/peps/pep-0484/) add static type annotations to Python. It would be nice for the Hypothesis API to support this, as the public API is actually pretty static in nature. stub files offer a compatibility mode, so this can be done transparently without dropping support for Python 2 (which is sadly there for the long haul).

Steps this would require:
1. Making Strategy a generic type with a single parameter, Strategy[T] that specifies the type of value generated. Note that Strategy is covariant in its type parameter.
2. Providing static types for all of the public API in hypothesis and hypothesis.strategies.
3. Running a mypy checker on Hypothesis + its test suite as part of the build

Things to note:
1. `given` and `builds` cannot be precisely typed with this type system, but they can still have types that are better than nothing by constraining their *args and **kwargs to have a strategy type (this is actually quite useful because it will catch two common gotchas when using this).
2. Settings is going to be quite hard to type correctly and is probably not worth bothering with.
3. I am _not_ proposing to make Hypothesis internals typed. In particular templates for strategies will remain completely untyped. I've worked out what typing this accurately requires in the past and PEP 0484 can't begin to express it.

### PR DESCRIPTION
**Pull request by Zac-HD**
This PR *finally* closes #200 :tada: 

For anyone manually exploring this PR, note that Mypy does *not* find hints installed in editable mode (`pip install -e .`); you have to do a straight `pip install .` - see python/mypy#5007 for work on a fix.  This does not affect our CI setup as we don't use editable mode there, but is annoying in local development.

### CODE REVIEW

**Comment by Zac-HD** [comment]
Pssst, @DRMacIver - any advice on a build setup for this?  All I want is an environment with both Mypy and Hypothesis installed...

**Comment by Zac-HD** [comment]
:tada:, I finally got the build working!  The solution is to set the `MYPYPATH` rather than relying on a PEP 561 install thing, but I'm OK with that as any breakage would be hugely obvious upstream before these tests gave a false-pass.

CC @DRMacIver for review; it's mostly a matter of documentation and what counts as public API I thing :smile:

**Comment by Zac-HD** [comment]
# Do not merge this yet

First, I need to:

- Report a minimal repro of the `one_of` problem upstream
- Retest this on latest Mypy
- Check that workarounds are still required (& note minimal mypy version in docs)
- Optionally add some more tests

**Comment by Zac-HD** [comment]
python/mypy#5218 is my report of (the first part I isolated of) the problem with `one_of`.  Once that's fixed, we might also be affected by python/mypy#3955, python/mypy#3644, and python/mypy#4432.

File: hypothesis-python/docs/details.rst [hunk 1]
```diff
@@ -543,3 +543,55 @@ changes between Python 3.5.0 and 3.6.1, including at minor versions.  These
...
+    Hypothesis' type hints may make breaking changes between minor releases.
+
+    Upstream tools and conventions about type hints remain in flux - for
+    example the :mod:`python:typing` module itself is provisional, and Mypy
+    has not yet reached version 1.0 - and we plan to support the latest
+    version of this ecosystem, as well as older versions where practical.
+
+    We may also find more precise ways to describe the type of various
+    interfaces, or change their type and runtime behaviour togther in a way
+    which is otherwise backwards-compatible.  We often ommit type hints for
```
Comment (alexwlchan) [inline hypothesis-python/docs/details.rst]: sp: “ommit”

File: whole-repo-tests/test_type_hints.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+def get_mypy_analysed_type(fname, val):
+    out = subprocess.Popen(
+        [tool_path('mypy'), fname],
+        stdout=subprocess.PIPE, encoding='utf-8', universal_newlines=True,
+        # We set the MYPYPATH explicitly, because PEP561 discovery wasn't
+        # working in CI as of mypy==0.600 - hopefully a temporary workaround.
+        env=dict(os.environ, MYPYPATH=tools.PYTHON_SRC),
+    ).stdout.read()
+    assert len(out.splitlines()) == 1
+    typ = out.split('error: Revealed type is ')[1].strip().strip("'")
```
Comment (alexwlchan) [inline whole-repo-tests/test_type_hints.py]: I have no idea what Mypy successful or unsuccessful output looks like – a brief comment here with examples of both would help satisfy me that this string parsing code is working correctly.

Comment (Zac-HD) [reply whole-repo-tests/test_type_hints.py]: I'll add a link to [the docs](https://mypy.readthedocs.io/en/latest/common_issues.html#displaying-the-type-of-an-expression) and an example of the output we're parsing.

File: whole-repo-tests/test_type_hints.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+
+@pytest.mark.parametrize('val,expect', [
+    ('integers()', 'int'),
+    ('text()', 'str'),
+    ('integers().map(str)', 'str'),
+    ('booleans().filter(bool)', 'bool'),
+    ('lists(none())', 'list[None]'),
+    ('dictionaries(integers(), datetimes())', 'dict[int, datetime.datetime]'),
+    ('recursive(integers(), lists)', 'Union[list[Ex`-1], int]'),
```
Comment (DRMacIver) [inline whole-repo-tests/test_type_hints.py]: What is ```Ex`-1```?

Comment (Zac-HD) [reply whole-repo-tests/test_type_hints.py]: `Ex` is the type of the example which `SearchStrategy` is generic and covariant in; and ```Ex`-1``` appears to denote recursion in this type - that is, ```Ex`-1``` == ```Union[list[Ex`-2], int]``` and so on.  
I'll add a comment to this effect...

File: hypothesis-python/setup.py [hunk 4]
```diff
@@ -86,7 +86,7 @@ def local_file(name):
     author_email='david@drmaciver.com',
     packages=setuptools.find_packages(SOURCE),
     package_dir={'': SOURCE},
-    # package_data={'': ['py.typed']},  # un-comment to release type hints
+    package_data={'': ['py.typed']},
```
Comment (ethanhs) [inline hypothesis-python/setup.py]: I believe it is better to explicitly mark the package name this file is from.

**Comment by Zac-HD** [comment]
✨ *Finally!* 🎉

**Comment by ethanhs** [comment]
FWIW, support for editable installs just landed and will be in the next mypy release.

**Pull request merged**